lib/tools/node-modules/node-modules-dependencies-builder.ts

```diff
@@ -0,0 +1,143 @@
+import * as path from "path";
+import * as fs from "fs";
+
+export class NodeModulesDependenciesBuilder implements INodeModulesDependenciesBuilder {
+    private projectPath: string;
+    private rootNodeModulesPath: string;
+    private resolvedDependencies: any[];
+    private seen: any;
+
+    public constructor(private $fs: IFileSystem) {
+        this.seen = {};
+        this.resolvedDependencies = [];
+    }
+
+    public getProductionDependencies(projectPath: string): any {
+        this.projectPath = projectPath;
+        this.rootNodeModulesPath = path.join(this.projectPath, "node_modules");
+
+        let projectPackageJsonpath = path.join(this.projectPath, "package.json");
+        let packageJsonContent = this.$fs.readJson(projectPackageJsonpath).wait();
+
+        _.keys(packageJsonContent.dependencies).forEach(dependencyName => {
+            let depth = 0;
+            let directory = path.join(this.rootNodeModulesPath, dependencyName);
+
+            // find and traverse child with name `key`, parent's directory -> dep.directory
+            this.traverseDependency(dependencyName, directory, depth);
+        });
+
+        return this.resolvedDependencies;
+    }
+
+    private traverseDependency(name: string, currentModulePath: string, depth: number): void {
+        // check if key appears in a scoped module dependency
+        let isScoped = name.indexOf('@') === 0;
+
+        if (!isScoped) {
```

Comment:
The scoped/unscoped branches seem to be mostly the same code copied/pasted over. Can we eliminate the duplication in a shared method and pass the different paths as parameters?